src/librustc_lint/builtin.rs

```diff
@@ -591,6 +614,28 @@ impl Deprecated {
             cx.span_lint(lint, span, &msg);
         }
     }
+
+    fn push_item(&mut self, cx: &LateContext, item_id: ast::NodeId) {
+        self.def_stack.push(cx.tcx.map.local_def_id(item_id).index);
+    }
+
+    fn item_post(&mut self, cx: &LateContext, item_id: ast::NodeId) {
+        let top = self.def_stack.pop();
+
+        if let Some(id) = top {
+            if DefId::local(id) == cx.tcx.map.local_def_id(item_id) { return; }
+        }
+
+        bug!("unexpected lint traversal order: expected  `_post` for {:?} but called for {:?}",
+             top, item_id);
+    }
```

Comment:
You mean we could use `get_parent` instead of a stack, and store a single `NodeId`?
I think that would work - the problem solved by the stack here is that not all `NodeId`s are in the map.